modules/porous_flow/src/materials/PorousFlowSimpleCO2.C

```diff
@@ -0,0 +1,562 @@
+  /// Density and derivatives wrt pressure and temperature at the qps
+  _density_qp[_qp] = density(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num]);
+  _ddensity_qp_dp[_qp] = dDensity_dP(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num]);
+  _ddensity_qp_dt[_qp] = dDensity_dT(_porepressure_qp[_qp][_phase_num], _temperature_qp[_qp][_phase_num]);
+
+  /// Viscosity and derivative wrt temperature at the nodes
+  _viscosity_nodal[_qp] = viscosity(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _density_nodal[_qp]);
+  _dviscosity_nodal_dt[_qp] = dViscosity_dT(_porepressure_nodal[_qp][_phase_num], _temperature_nodal[_qp][_phase_num], _density_nodal[_qp]);
+}
+
+Real
+PorousFlowSimpleCO2::density(Real pressure, Real temperature) const
+{
+  Real rho;
+
+  if (pressure <= _p_critical)
+    rho = gasDensity(pressure, temperature);
+
+  else // if (pressure > _p_critical)
+    rho = supercriticalDensity(pressure, temperature);
+
+  return rho;
+}
+
+Real
+PorousFlowSimpleCO2::viscosity(Real pressure, Real temperature, Real density) const
+{
+  Real mu;
+
+  if (pressure <= _p_critical)
+    mu = gasViscosity(temperature, density);
+
+  else // if (pressure > _p_critical)
+    mu = supercriticalViscosity(pressure, temperature);
+
+  return mu;
+}
+
+Real
+PorousFlowSimpleCO2::dDensity_dP(Real pressure, Real temperature) const
+{
+  Real drho;
+
+  if (pressure <= _p_critical)
+    drho = dGasDensity_dP(pressure, temperature);
+
+  else // if (pressure > _p_critical)
+    drho = dSupercriticalDensity_dP(pressure, temperature);
+
+  return drho;
+}
+
+Real
+PorousFlowSimpleCO2::dDensity_dT(Real pressure, Real temperature) const
+{
+  Real drho;
+
+  if (pressure <= _p_critical)
+    drho = dGasDensity_dT(pressure, temperature);
+
+  else // if (pressure > _p_critical)
+    drho = dSupercriticalDensity_dT(pressure, temperature);
+
+  return drho;
+}
+
+Real
+PorousFlowSimpleCO2::gasViscosity(Real temperature, Real density) const
+{
+  Real tk = temperature + _t_c2k;
+  Real tstar = tk / 251.196;
+  Real a[5] = {0.235156, -0.491266, 5.211155e-2, 5.347906e-2, -1.537102e-2};
+  Real d[5] = {0.4071119e-2, 0.7198037e-4, 0.2411697e-16, 0.2971072e-22, -0.1627888e-22};
+  int j[5] = {1, 1, 4, 1, 2};
+  int i[5] = {1, 2, 6, 8, 8};
+
+  // Zero-denisty viscosity
+  Real sum = 0.0;
+
+  for (int n = 0; n < 5; ++n)
+    sum += a[n] * std::pow(std::log(tstar), n);
+
+  Real theta = std::exp(sum);
+  Real mu0 = 1.00697 * std::sqrt(tk) / theta;
+
+  Real b[5];
+  for (unsigned int n = 0; n < 5; ++n)
+    b[n] = d[n] /  std::pow(tstar, j[n] - 1);
+
+  // Excess viscosity due to density
+  Real mu = 0.0;
+
+  for (unsigned int n = 0; n < 5; ++n)
+    mu += b[n] * std::pow(density, i[n]);
+
+  return (mu0 + mu) * 1e-6; // convert to Pa.s
+}
+
+Real
+PorousFlowSimpleCO2::supercriticalViscosity(Real pressure, Real temperature) const
+{
+  Real b0[5] = {-1.958098980443E+01, 1.123243298270E+00, -2.320378874100E-02,
```

Comment:
those should all be const. Also a literature reference as a comment might be helpful.